src/abstractdataframe/nest.jl

```diff
@@ -0,0 +1,390 @@
+                               for (n, c) in pairs(eachcol(v))],
+                              copycols=false) |> only
+            elseif Tables.istable(v) # produce NamedTuple
+                v = Tables.columntable(v)
+            end
+            push!(tmp_df, v, cols=cols, promote=promote)
+        end
+        hcat!(ref_df, tmp_df, makeunique=makeunique, copycols=false)
+    end
+    return if flatten
+        DataFrames.flatten(ref_df, col_count+1:ncol(ref_df), scalar=Missing)
+    else
+        ref_df
+    end
+end
+
+"""
+    flatten(df::AbstractDataFrame, cols; scalar::Type)
+
+When columns `cols` of data frame `df` have iterable elements that define
+`length` (for example a `Vector` of `Vector`s), return a `DataFrame` where each
+element of each `col` in `cols` is flattened, meaning the column corresponding
+to `col` becomes a longer vector where the original entries are concatenated.
+Elements of row `i` of `df` in columns other than `cols` will be repeated
+according to the length of `df[i, col]`. These lengths must therefore be the
+same for each `col` in `cols`, or else an error is raised. Note that these
+elements are not copied, and thus if they are mutable changing them in the
+returned `DataFrame` will affect `df`.
+
+`cols` can be any column selector ($COLUMNINDEX_STR; $MULTICOLUMNINDEX_STR).
+
+If `scalar` is passed then values that have this type in flattened columns
+are treated as scalars and broadcasted as many times as is needed to match
+lengths of values stored in other columns. One row is produced if all
+corresponding values are scalars.
+
+$METADATA_FIXED
+
+# Examples
+
+```jldoctest
+julia> df1 = DataFrame(a=[1, 2], b=[[1, 2], [3, 4]], c=[[5, 6], [7, 8]])
+2×3 DataFrame
+ Row │ a      b       c
+     │ Int64  Array…  Array…
+─────┼───────────────────────
+   1 │     1  [1, 2]  [5, 6]
+   2 │     2  [3, 4]  [7, 8]
+
+julia> flatten(df1, :b)
+4×3 DataFrame
+ Row │ a      b      c
+     │ Int64  Int64  Array…
+─────┼──────────────────────
+   1 │     1      1  [5, 6]
+   2 │     1      2  [5, 6]
+   3 │     2      3  [7, 8]
+   4 │     2      4  [7, 8]
+
+julia> flatten(df1, [:b, :c])
+4×3 DataFrame
+ Row │ a      b      c
+     │ Int64  Int64  Int64
+─────┼─────────────────────
+   1 │     1      1      5
+   2 │     1      2      6
+   3 │     2      3      7
+   4 │     2      4      8
+
+julia> df2 = DataFrame(a=[1, 2], b=[("p", "q"), ("r", "s")])
+2×2 DataFrame
+ Row │ a      b
+     │ Int64  Tuple…
+─────┼───────────────────
+   1 │     1  ("p", "q")
+   2 │     2  ("r", "s")
+
+julia> flatten(df2, :b)
+4×2 DataFrame
+ Row │ a      b
+     │ Int64  String
+─────┼───────────────
+   1 │     1  p
+   2 │     1  q
+   3 │     2  r
+   4 │     2  s
+
+julia> df3 = DataFrame(a=[1, 2], b=[[1, 2], [3, 4]], c=[[5, 6], [7]])
+2×3 DataFrame
+ Row │ a      b       c
+     │ Int64  Array…  Array…
+─────┼───────────────────────
+   1 │     1  [1, 2]  [5, 6]
+   2 │     2  [3, 4]  [7]
+
+julia> flatten(df3, [:b, :c])
```

Comment:
Is `equal-length-zip` the desired behavior here? The alternatives that come to mind are `zip-shortest`, `zip-longest`, and especially `cartesian-product`.

Reply:
This is what is the current behavior. We could add a kwarg for other behaviors (but I would discuss them in a separate issue, for creating cartesian product we have `allcombinations`).